File: bto_management_system/service/report_service.py
from typing import List, Dict, Optional
from .interfaces.ireport_service import IReportService
from .interfaces.iproject_service import IProjectService # Use interface
from repository.interfaces.iapplication_repository import IApplicationRepository
from repository.interfaces.iuser_repository import IUserRepository
from common.enums import ApplicationStatus, FlatType

class ReportService(IReportService):
    """Handles generation of reports."""
    def __init__(self, application_repository: IApplicationRepository,
                 project_service: IProjectService,
                 user_repository: IUserRepository):
        self._app_repo = application_repository
        self._project_service = project_service
        self._user_repo = user_repository

    def generate_booking_report_data(self, filter_project_name: Optional[str] = None,
                                     filter_flat_type_str: Optional[str] = None,
                                     filter_marital: Optional[str] = None) -> List[Dict]:
        """Generates data for the booking report based on filters."""
        report_data = []
        booked_apps = [app for app in self._app_repo.get_all() if app.status == ApplicationStatus.BOOKED]

        filter_flat_type: Optional[FlatType] = None
        if filter_flat_type_str:
            try: filter_flat_type = FlatType.from_value(filter_flat_type_str)
            except ValueError: pass # Ignore invalid filter

        filter_marital_lower = filter_marital.lower() if filter_marital else None

        for app in booked_apps:
            project = self._project_service.find_project_by_name(app.project_name)
            applicant = self._user_repo.find_user_by_nric(app.applicant_nric)

            if not project or not applicant:
                print(f"Warning: Skipping report entry for application {app.applicant_nric}-{app.project_name} due to missing project/applicant.")
                continue

            # Apply filters
            if filter_project_name and project.project_name.lower() != filter_project_name.lower(): continue
            if filter_flat_type and app.flat_type != filter_flat_type: continue
            if filter_marital_lower and applicant.marital_status.lower() != filter_marital_lower: continue

            report_data.append({
                "NRIC": app.applicant_nric,
                "Applicant Name": applicant.name,
                "Age": applicant.age,
                "Marital Status": applicant.marital_status,
                "Flat Type": app.flat_type.to_string(),
                "Project Name": project.project_name,
                "Neighborhood": project.neighborhood
            })

        # Sort report data
        report_data.sort(key=lambda x: (x["Project Name"], x["Applicant Name"]))
        return report_data

File: J2/Services/ReportService.java
package Services;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import Enums.ApplicationStatus;
import Enums.FlatType;
import Enums.MaritalStatus;
import Interfaces.Repositories.IApplicationRepository;
import Interfaces.Repositories.IUserRepository;
import Interfaces.Services.IReportService;
import Models.BTOApplication;
import Models.User;

public class ReportService implements IReportService {

    private final IApplicationRepository applicationRepository;
    private final IUserRepository userRepository;

    public ReportService(IApplicationRepository applicationRepository, IUserRepository userRepository) {
        this.applicationRepository = applicationRepository;
        this.userRepository = userRepository;
    }

    @Override
    public List<BTOApplication> generateBookedApplicantReport(List<String> projectNames, FlatType filterFlatType,
                                                              MaritalStatus filterMaritalStatus, int minAge, int maxAge) {

        if (projectNames == null || projectNames.isEmpty()) {
            System.out.println("No projects specified for the report.");
            return List.of(); // Return empty list
        }

        // Fetch all applications once
        Map<String, BTOApplication> allApplications = applicationRepository.getAllApplications();
        Map<String, User> allUsers = userRepository.getAllUsers(); // Fetch users for filtering

        return allApplications.values().stream()
                // Filter by BOOKED status
                .filter(app -> app.getStatus() == ApplicationStatus.BOOKED)
                // Filter by specified project names
                .filter(app -> projectNames.contains(app.getProjectName()))
                // Filter by flat type (if specified)
                .filter(app -> filterFlatType == null || app.getFlatTypeApplied() == filterFlatType)
                // Filter by applicant details (marital status, age)
                .filter(app -> {
                    User user = allUsers.get(app.getApplicantNric());
                    if (user == null) return false; // Skip if user data not found

                    // Apply marital status filter
                    if (filterMaritalStatus != null && user.getMaritalStatus() != filterMaritalStatus) {
                        return false;
                    }
                    // Apply min age filter (maxAge=0 means no upper limit)
                    if (minAge > 0 && user.getAge() < minAge) {
                        return false;
                    }
                    // Apply max age filter (maxAge=0 means no upper limit)
                    if (maxAge > 0 && user.getAge() > maxAge) {
                        return false;
                    }
                    return true; // Passed all filters
                })
                // Sort the results
                .sorted(Comparator.comparing(BTOApplication::getProjectName)
                                  .thenComparing(BTOApplication::getApplicantNric))
                .collect(Collectors.toList());
    }

    @Override
    public void displayApplicantReport(List<BTOApplication> reportData, FlatType filterFlatType,
                                       MaritalStatus filterMaritalStatus, int minAge, int maxAge) {

        System.out.println("\n--- Report: Applicants with Flat Bookings ---");
        System.out.println("Filters Applied: Project(s) selected, FlatType="
                + (filterFlatType == null ? "Any" : filterFlatType.getDisplayName()) + ", MaritalStatus="
                + (filterMaritalStatus == null ? "Any" : filterMaritalStatus) + ", Age="
                + (minAge > 0 ? minAge : "Any") + "-" + (maxAge > 0 ? maxAge : "Any"));
        System.out.println("---------------------------------------------------------------------------------");
        System.out.printf("%-15s | %-15s | %-5s | %-10s | %-15s | %-8s\n",
                "Applicant NRIC", "Name", "Age", "Marital", "Project Name", "FlatType");
        System.out.println("---------------------------------------------------------------------------------");

        if (reportData == null || reportData.isEmpty()) {
            System.out.println("No matching booked applications found for the specified filters.");
        } else {
            // Fetch all users once to avoid repeated lookups inside the loop
            Map<String, User> allUsers = userRepository.getAllUsers();
            reportData.forEach(app -> {
                User user = allUsers.get(app.getApplicantNric());
                System.out.printf("%-15s | %-15s | %-5s | %-10s | %-15s | %-8s\n",
                        app.getApplicantNric(),
                        (user != null) ? user.getName() : "N/A",
                        (user != null) ? String.valueOf(user.getAge()) : "N/A", // Age as string
                        (user != null) ? user.getMaritalStatus().name() : "N/A",
                        app.getProjectName(),
                        (app.getFlatTypeApplied() != null) ? app.getFlatTypeApplied().getDisplayName() : "N/A");
            });
        }
        System.out.println("---------------------------------------------------------------------------------");
        System.out.println("Total matching booked applicants: " + (reportData == null ? 0 : reportData.size()));
        System.out.println("--- End of Report ---");
    }
}
